gui/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,615 @@
+
+from typing import Union
+
+import numpy as np
+
+from .._base_layer import Layer
+from .._register import add_to_viewer
+from ..._vispy.scene.visuals import Line as LinesNode
+from vispy.color import get_color_names
+
+from vispy.util.event import Event
+
+from .view import QtVectorsLayer
+from .view import QtVectorsControls
+
+
+@add_to_viewer
+class Vectors(Layer):
+    """
+    Vectors layer renders lines onto the image.
+    There are currently two data modes supported:
+        1) image-like vectors where every pixel defines its own vector
+        2) coordinate-like vectors where x-y position is not fixed to a grid
+    Supports ONLY 2d vector field
+
+    Properties
+    ----------
+    vectors : np.ndarray
+        array of shape (N,4) or (N, M , 2)
+
+    data : np.ndarray
+        ABC implementation.
+        returns vectors property setter/getter
+
+    _original_data : np.ndarray
+        Used by averaging and length adjustments.
+        Is set during layer construction.
+        Is NOT updated if data is adjusted by assigning the 'vectors' property
+
+    averaging : str
+        kernel over which to average from one of _kernel_dict
+        averaging.setter adjusts the underlying data
+        subscribe an observer by registering it with "averaging_bind_to"
+
+    width : int
+        width of the line in pixels
+
+    length : int or float
+        length of the line
+        length.setter adjusts the underlying data
+        subscribe an observer by registering it with "length_bind_to"
+
+    color : str
+        one of "get_color_names" from vispy.color
+
+    conector : str or np.ndarray
+        way to render line between coordinates
+        two built in types: "segment" or "strip"
+        third type is np.ndarray
+
+    mode : str
+        control panel mode
+
+    Attributes
+    ----------
+    Private attributes
+    -------
+    _kernel_dict
+    _kernel
+    _data_types
+    _data_type
+    _width
+    _color
+    _colors
+    _connector_types
+    _connector
+    _avg_dims
+    _averaging
+    _length
+    _avg_observers
```

Comment:
Why do you have the `_avg_observers` and `_len_observers`. Could this be accomplished with the event based implementation that we put on properties in #155. In this case setting the `averaging` or `length` properties would trigger the `averaging` or `length` events that could be hooked up to whatever other things you wanted

Reply:
Yes this is totally possible.  I remember Loic also mentioning a listener module at the last meeting.  Events could do the trick too.

The reason these methods exist is because they adjust the underlying data.  The literal way i've done vector averaging (in _default_avg and _default_length) doesn't always make sense depending on the application/data.  Registering an observer allows one to pay attention to user-GUI inputs and have external code adjust the data in another, meaningful way.

Reply:
removed this observer pattern in favor of vispy's event system.  There is no longer any tracking of observers in a list within the vector_layer.

I moved the averaging and length adjustment math to the qt properties, but am open to switching this.  My reasoning is that qt properties contains slots, while the layer appears to contain emitters, and the averaging/length functions behave mostly as a slot.